### PR TITLE
Folder metadata index

### PR DESCRIPTION
**Pull request by YurkoWasHere**

### CODE REVIEW

File: contrib/sideload_csv2json.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
+# sideload_csv2json
+
+This script will convert a CSV file into a JSON index used by the preprocessor with configuration set in the `config.json` file
+
+
+## .config.json
```
Comment (makeworld-the-better-one) [inline contrib/sideload_csv2json.md:6]: It's unclear why this starts with a `.`, same with the other heading.

File: contrib/sideload_csv2json.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+startrow = 0
+endrow = 1000000000
+filename = "Starling-Caption-Credit.csv"
+config_filename = "Starling-Caption-Credit-config.json"
+
+# Load config
+with open(config_filename, "r") as config_file:
+    config = json.load(config_file)
+
+with open(filename, newline="\n", encoding="utf8") as csvfile:
```
Comment (makeworld-the-better-one) [inline contrib/sideload_csv2json.py:15]:
```suggestion
with open(filename, newline="", encoding="utf8") as csvfile:
```

`newline=""` is required according to the csv library docs: https://docs.python.org/3/library/csv.html#csv.reader

File: contrib/sideload_csv2json.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            heading = row
+        else:
+
+            countlines = countlines + 1
+
+            json_metadata = deepcopy(json_metadata_template)
+            column_index = 0
+            for item in row:
+                if len(item) > 1 and item[1] == "[":
+                    json_metadata[heading[column_index]] = ast.literal_eval(item)
```
Comment (makeworld-the-better-one) [inline contrib/sideload_csv2json.py:44]: Maybe wrap this in a try-except block so you still ingest values even if they can't be evaluated. Like `[foo, bar]`.

File: contrib/sideload_csv2json.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+                if col_name == "":
+                    col_name = "col_" + str(column_index)
+                json_metadata_template[col_name] = ""
+                column_index += 1
+            heading = row
+        else:
+
+            countlines = countlines + 1
+
+            json_metadata = deepcopy(json_metadata_template)
```
Comment (makeworld-the-better-one) [inline contrib/sideload_csv2json.py:40]: Right before you process anything, you should check whether `countlines` is equal to or greater than `startrow`, and just `continue` if it's not. Right now you still process rows even if you don't add them in the end, which is slower.

File: folder/README.md [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+sourcePath: path to watch for new files
+targetPath: Path to place ready bundles into
+allowedPatterns: Extensions to watch
+extractName: Extract the name from the filename
+extractNameCharacters: Delimiter for name
+method: Note about how the file arrived
+processWacz: Process metadata content from wacz files
+processProofMode" Process metadata content from proofmode
+lockFile: Wait for file not to exist before processing file
+author: Author Metadata
```
Comment (makeworld-the-better-one) [inline folder/README.md:16]: Markdown places all of these on the same line, making it an unreadable paragraph. Put them as bullet points instead.

File: folder/main.py [hunk 6]
```diff
@@ -112,13 +158,31 @@ def __init__(self, conf):
...
+        if os.path.basename(event.src_path) == "index.json":
+            logging.info(f"Skipping index.json file")
+            return
+
         logging.info(f"Starting Processing of file {event.src_path}")
         sha256asset = common.sha256sum(event.src_path)
 
+        lock_file = "" 
+        if "lockFile" in self.config:
+            lock_file = self.config["lockFile"]
```
Comment (makeworld-the-better-one) [inline folder/main.py:172]: This field is not mentioned in the docs.

**Review by makeworld-the-better-one — APPROVED** [review]
If we have to deal with huge CSV files, it might become better to write a json lines format, so only one row has to be held in memory at a time. For now it doesn't seem like a concern.

File: folder/main.py (hunk 6 above)
Comment (YurkoWasHere) [reply folder/main.py:172]: I think it does?
```
- lockFile: Wait for file not to exist before processing file
```

File: contrib/sideload_csv2json.py (hunk 3 above)
Comment (YurkoWasHere) [reply contrib/sideload_csv2json.py:44]:
```suggestion
                    try:
                        json_metadata[heading[column_index]] = ast.literal_eval(item)
                    except:
                        json_metadata[heading[column_index]] = item
```

File: contrib/sideload_csv2json.py (hunk 4 above)
Comment (YurkoWasHere) [inline contrib/sideload_csv2json.py:40]:
```suggestion
            if countlines >= startrow:
                continue
            json_metadata = deepcopy(json_metadata_template)
```

**Pull request merged**